GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -340,28 +365,18 @@ private void WriteEvent(string eventName, EventLevel level, Keywords keywords, E
                 Payload = jsonPayload
             };
 
+            // Iterating over the bag is thread-safe as the enumerator returned here
+            // is of a snapshot of the bag.
             foreach (EventListener listener in this.listeners)
             {
-                string errorMessage;
-                bool? success = listener.TryRecordMessage(message, out errorMessage);
-
-                // Try to mark success or failure for this listener only if it was enabled for this message type
-                if (success.HasValue)
-                {
-                    if (success == true)
-                    {
-                        this.MarkAndLogListenerRecovery(listener);
-                    }
-                    else
-                    {
-                        this.MarkAndLogListenerFailure(listener, errorMessage);
-                    }
-                }
+                listener.RecordMessage(message);

```

Comment:
I have mixed opinions on using events to signal that a problem occurred... I liked how the previous pattern handled these errors inline in the the caller, or it least it seemed simpler to follow then the event pattern. As it is, are there other subscribers to the success / failure events?

(I do like how the new pattern avoids the `TryXXX` pattern for this method, as we expect it to generally succeed).

Reply:
Do you have another suggestion how to report failures asynchronously/from the other thread/detached from the initial method call that enqueued the data?

Reply:
Could the caller collect and handle the failures?

Maybe I missed it (I will try and look again when I am back at computer), but was there a deficiency in the original pattern where this was handled in the loop that delivered messages to the listeners?

Reply:
That write loop is still happening, but now it's invoked from the 'failure' event.
Previously it came in from the _return value_ of the `TryRecordMessage` call.. But since we don't want to hold this call up, we don't know the outcome of the invocation until some indeterminate time later.

Reply:
Thinking about this and another option would be for the JsonTracer to implement an interface like:
```
public interface IListenerHealthHandler
{
    void Recovered(EventListener listener);
    void Failed(EventListener listener, string errorMessage);
}
```

And set that on each listener.  Pretty much the same as an event handler but more explicit and there could be only one.

Reply:
> But since we don't want to hold this call up, we don't know the outcome of the invocation until some indeterminate time later.

Thanks for the clarification!

Reply:
@kewillford I've updated the PR to replace events with an event sink object & interface as suggested. I went with the term 'event sink' rather than 'handler'.